iib/workers/tasks/build.py

```diff
@@ -80,6 +200,151 @@ def _get_local_pull_spec(request_id):
+    if skopeo_raw['mediaType'] != 'application/vnd.docker.distribution.manifest.list.v2+json':
+        raise IIBError(f'The pull specification of {pull_spec} is not a v2 manifest list')
+
+    arches = set()
+    for manifest in skopeo_raw['manifests']:
+        arches.add(manifest['platform']['architecture'])
+
+    return arches
+
+
+def _get_resolved_image(pull_spec):
+    """
+    Get the pull specification of the image using its digest.
+
+    :param str pull_spec: the pull specification of the image to resolve
+    :return: the resolved pull specification
+    :rtype: str
+    """
+    log.debug('Resolving %s', pull_spec)
+    skopeo_output = _skopeo_inspect(f'docker://{pull_spec}')
+    pull_spec_resolved = f'{skopeo_output["Name"]}@{skopeo_output["Digest"]}'
+    log.debug('%s resolved to %s', pull_spec, pull_spec_resolved)
+    return pull_spec_resolved
+
+
+def _prepare_request_for_build(binary_image, request_id, from_index=None, add_arches=None):
+    """
+    Prepare the request for the architecture specific builds.
+
+    All information that was retrieved and/or calculated for the next steps in the build are
+    returned as a dictionary.
+
+    This function was created so that code didn't need to be duplicated for the ``add`` and ``rm``
+    request types.
+
+    :param str binary_image: the pull specification of the image where the opm binary gets copied
+        from.
+    :param int request_id: the ID of the IIB build request
+    :param str from_index: the pull specification of the image containing the index that the index
+        image build will be based from.
+    :return: a dictionary with the keys: arches, binary_image_resolved, and fom_index_resolved.
+    :raises iib.exceptions.IIBError: if the image resolution fails or the architectures couldn't
+        be detected.
+    """
+    set_request_state(request_id, 'in_progress', 'Resolving the images')
+
+    if add_arches:
+        arches = set(add_arches)
+    else:
+        arches = set()
+
+    binary_image_resolved = _get_resolved_image(binary_image)
+    binary_image_arches = _get_image_arches(binary_image_resolved)
+
+    if from_index:
+        from_index_resolved = _get_resolved_image(from_index)
+        from_index_arches = _get_image_arches(from_index_resolved)
+        arches = arches | from_index_arches
+    else:
+        from_index_resolved = None
+
+    if not arches:
+        raise IIBError('No arches were provided to build the index image')
+
+    arches_str = ', '.join(sorted(arches))
+    log.debug('Set to build the index image for the following arches: %s', arches_str)
+
+    if not arches.issubset(binary_image_arches):
+        raise IIBError(
+            'The binary image is not available for the following arches: {}'.format(
+                ', '.join(sorted(arches - binary_image_arches))
+            )
+        )
+
+    conf = get_worker_config()
+    if not arches.issubset(conf['iib_arches']):
```

Comment:
Since we have this, do you think we should remove s390x from our list for now? Since we won't be supporting it when we first deploy IIB?